recipes-extended/xen/files/libxl-blktap3-support.patch

```diff
@@ -0,0 +1,703 @@
++    INIT_LIST_HEAD(&list);
++
++    if ((err = tap_ctl_list(&list)))
++        return err;
++
++    /* TAILQ_EMPTY(&list)--> old function */
++    if (list_empty(&list))
++        return ret;
++
++    tap_list_for_each_entry_safe(entry, next_t, &list) {
++
++        if (type && (!entry->type || strcmp(entry->type, type)))
++            continue;
++
++        if (path && (!entry->path || strcmp(entry->path, path)))
++            continue;
++
++        *tap = *entry;
++	tap->type = tap->path = NULL;
++        ret = 0;
++        break;
++    }
++
++    tap_ctl_list_free(&list);
++
++    return ret;
++}
++
++/**
++ * blktap3 doesn't require blkback, so it's always available.
++ */
++int libxl__blktap_enabled(libxl__gc *gc)
++{
++    return 1;
++}
++
++char *libxl__blktap_devpath(libxl__gc *gc, const char *disk,
++		libxl_disk_format format)
++{
++    const char *type = NULL;
++    char *params, *devname = NULL;
++    struct tap_list tap;
++    int err = 0;
++    int flags = 0;
++
++    type = libxl__device_disk_string_of_format(format);
++    if (!(err = blktap_find(type, disk, &tap))) {
++        LOG(DEBUG, "found tapdisk\n");
++	devname = libxl__sprintf(gc, "/dev/xen/blktap-2/tapdev%d", tap.minor);
++        if (devname)
++            return devname;
++		/* return 0; revert-back-changes*/
++    }
++
++    LOG(DEBUG, "tapdisk not found\n");
++
++	/* TODO Should we worry about return codes other than ENOENT? */
++
++    params = libxl__sprintf(gc, "%s:%s", type, disk);
++
++    /* tap_ctl_create(params, 0, -1, NULL) --> old function call  */
++    if (!(err = tap_ctl_create(params, &devname, flags, -1, 0, 0))) {
++        LOG(DEBUG, "created tapdisk\n");
++        return devname; /*revert-back-changes*/
++    }
++
++    LOG(ERROR, "error creating tapdisk: %s\n", strerror(err));
++
++    return NULL;/*revert-back-changes*/
++}
++
++int libxl__device_destroy_tapdisk(libxl__gc *gc, const char *be_path)
++{
++    char *disk;
++    int err;
++    struct tap_list tap;
++
++	LOG(DEBUG, "destroying tapdisk %s\n", be_path);
++
++    disk = strchr(be_path, ':');
++    if (!disk) {
++        LOG(ERROR, "Unable to parse params %s", be_path);
++        return ERROR_INVAL;
++    }
++
++    *disk++ = '\0';
++
++    err = blktap_find(be_path, disk, &tap);
++    if (err < 0) {
++        /* returns -errno */
++        LOGEV(ERROR, -err, "Unable to find type %s disk %s", be_path, disk);
++        return ERROR_FAIL;
++    }
++
++    err = tap_ctl_destroy(tap.pid, tap.minor, 0, NULL);
++    if (err < 0) {
++        LOGEV(ERROR, -err, "Failed to destroy tap device id %d minor %d",
++              tap.pid, tap.minor);
++        return ERROR_FAIL;
++    }
++
++    return 0;
++}
++
++/*
++ * Local variables:
++ * mode: C
++ * c-basic-offset: 4
++ * indent-tabs-mode: nil
++ * End:
++ */
+Index: xen-4.9.1/tools/libxl/list.h
+===================================================================
+--- /dev/null
++++ xen-4.9.1/tools/libxl/list.h
```

Comment:
Libxl already has a list implementation, _libxl_list.h, it's autogenerated at compile time.  I would imagine that should be sufficient for whatever blktap3 needs it for.  No need to grab this from NetBSD.

Take a look at LIBXL_LIST_HEAD, LIBXL_LIST_HEAD_INITIALIZER, etc.

Reply:
This was added as part of the base libxl patch, check here: https://github.com/qiaohaiming/blktap3/blob/master/libxl.diff, the list API's are conforming to the list.h added as part of this patch.  As suggested, the libxl list implementation might be sufficient but this will require changing the API's and list structure used for blktap3 support.

Reply:
Yeah, the lists are shared with the tap-ctl lib from blktap3.git, so they need to use the same list.h...
I suggest keeping stuff as is, but renaming list.h to tap-ctl-list.h to make things more obvious.
@crogers1 how does that sound to you?

Reply:
Fixed, no list file created anymore.